core/src/main/java/zipkin2/reporter/internal/CountBoundedQueue.java

```diff
@@ -0,0 +1,121 @@
+ * Copyright The OpenZipkin Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package zipkin2.reporter.internal;
+
+import java.util.Arrays;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+import zipkin2.reporter.ReporterMetrics;
+
+/**
+ * Multi-producer, multi-consumer queue that is bounded by count.
+ *
+ * <p>This is similar to {@link java.util.concurrent.ArrayBlockingQueue} in implementation.
+ */
+final class CountBoundedQueue<S> extends BoundedQueue<S> {
+
+  final ReentrantLock lock = new ReentrantLock(false);
+  final Condition available = lock.newCondition();
+  final ReporterMetrics metrics;
+
+  final int maxSize;
+
+  final S[] elements;
+  int count;
+  int writePos;
+  int readPos;
+
+  @SuppressWarnings("unchecked") CountBoundedQueue(ReporterMetrics metrics, int maxSize) {
+    this.metrics = metrics;
+    this.elements = (S[]) new Object[maxSize];
+    this.maxSize = maxSize;
+  }
+
+  @Override public boolean offer(S next, int nextSizeInBytes) {
+    return offer(next);
+  }
+
+  /**
+   * Returns true if the element could be added or false if it could not due to its size.
+   */
+  @Override public boolean offer(S next) {
+    lock.lock();
+    try {
+      if (count == maxSize) return false;
+
+      elements[writePos++] = next;
+
+      if (writePos == maxSize) writePos = 0; // circle back to the front of the array
+
+      count++;
+
+      available.signal(); // alert any drainers
+      return true;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /** Blocks for up to nanosTimeout for spans to appear. Then, consume as many as possible. */
+  @Override int drainTo(SpanWithSizeConsumer<S> consumer, long nanosTimeout) {
+    try {
+      // This may be called by multiple threads. If one is holding a lock, another is waiting. We
+      // use lockInterruptibly to ensure the one waiting can be interrupted.
+      lock.lockInterruptibly();
+      try {
+        long nanosLeft = nanosTimeout;
+        while (count == 0) {
+          if (nanosLeft <= 0) return 0;
+          nanosLeft = available.awaitNanos(nanosLeft);
+        }
+        return doDrain(consumer);
+      } finally {
+        lock.unlock();
+      }
+    } catch (InterruptedException e) {
+      return 0;
+    } finally {
+        // record after draining reduces the amount of gauge events vs on doing this on report
+        metrics.updateQueuedSpans(count);
+    }
+  }
+
+  /** Clears the queue unconditionally and returns count of spans cleared. */
+  @Override public int clear() {
+    lock.lock();
+    try {
+      int result = count;
+      count = readPos = writePos = 0;
+      Arrays.fill(elements, null);
+      return result;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  int doDrain(SpanWithSizeConsumer<S> consumer) {
+    int drainedCount = 0;
+    while (drainedCount < count) {
+      S next = elements[readPos];
+
+      if (next == null) break;
+      if (consumer.offer(next, 0)) {
```

Comment:
this is a problem area. So, we have messageMaxBytes, but we aren't checking and passing zero anyway.

We still need logic like this here:
```
      int nextSizeInBytes = encoder.sizeInBytes(next);
      int messageSizeOfNextSpan = sender.messageSizeInBytes(nextSizeInBytes);
      metrics.incrementSpanBytes(nextSizeInBytes);
      // don't encoding something larger than we can send
      if (messageSizeOfNextSpan > messageMaxBytes) // do something
```

Reply:
So the main thing we are doing with count bounded, is moving this calculation to the async reporter thread. We can no longer guard prior to enqueuing a span, but we still need to make sure we don't OOM or encode something larger than our max size. make sense?

Reply:
> So the main thing we are doing with count bounded, is moving this calculation to the async reporter thread. 

Correct

> We can no longer guard prior to enqueuing a span, but we still need to make sure we don't OOM or encode something larger than our max size. make sense?

So here is the trade-off: if we bring the message size calculation back, we will be losing all gains in `report` for this queue. There is a risk of OOM, it is true

Reply:
I was under the impression we were trying to solve getting the expense off the critical path, which does not require us to be careless about OOM.

Reply:
the expense we are talking about is in this thread now

```
    @Override public void run() {
      try {
        while (!result.closed.get()) {
          result.flush(consumer); // <-- this is where we are talking about the 1us size calc
        }
```

Reply:
> I was under the impression we were trying to solve getting the expense off the critical path, which does not require us to be careless about OOM. 

Hm ... the flush was non issue (it is always separate thread), but `report` was (https://github.com/openzipkin/brave/issues/1426):

```
    zipkin2.reporter.ByteBoundedQueue.offer(ByteBoundedQueue.java:51)
    zipkin2.reporter.AsyncReporter$BoundedAsyncReporter.report(AsyncReporter.java:259)
```

I think if we bring this checks back than we don't really need `CountBoundedQueue` ... But I am all ears here - let's do the right thing